src/projects.h

```diff
@@ -254,6 +255,60 @@ struct PJconsts {
+
+                              E R R O R   R E P O R T I N G
+
+    **************************************************************************************
+
+        Currently, we're doing error reporting through the context->last_errno indicator.
+
+        It is, however, not entirely sure this will be the right way to do it in all
+        cases: During allocation/initialization, it is certainly nice to have a higher
+        level error indicator, since we primarily signal "something went wrong", by
+        returning 0 from the pj_init family of functions - and with a null return we
+        cannot pass messages through internal state in the PJ object.
+
+        Historically, the errno variable has been used for that kind of messages, but
+        apparently, thread safety was added to PROJ.4 at a time where it was not clear
+        that errno is actually thread local.
+
+        Additionally, errno semantics has historically been misinterpreted in parts of
+        pj_init.c, a misinterpretation, that was mitigated by a hack in pj_malloc.c
+        some 15 years ago.
+
+        This PJ-local errno is a first step towards a more structured approach to
+        error reporting, being implemented in the OBS_API (cf. pj_obs_api.[ch]), and
+        related plumbing efforts.
+        
+        In due course this will let us get rid of the pj_malloc.c hack, and allow us
+        to introduce a more layered error reporting structure, where errors are
+        reported where they occur, and bubble up to the higher levels (context->errno,
+        then thread local errno), so the highest level indicate "something went wrong
+        somewhere", then the more localized ones can be used for pinpointing: 
+        
+        errno:               "something went wrong somewhere on this thread",
+        context->last_errno: "something went wrong in some PROJ.4 related code",
+        PJ->last_errno:      "It was in THIS PJ something went wrong",
+        pj_strerrno:         "This was what went wrong".
+
+        Which will be quite helpful, once fully implemented, especially for 
+        debugging complex transformation pipelines, while still maintaining backward
+        compatibility in the messaging system.
+
+        Note that there is even a global pj_errno, which is here and there accessed
+        without acquiring lock. This, and the practise of resetting the thread local
+        errno, should be given some consideration during the cleanup of the error
+        reporting system.
+
+        The name "last_errno", rather than "errno" is used partially for alignment
+        with the context->last_errno, partially because in a multithreaded environment,
+        errno is a macro, and spurious spaces turning "errno" into a separate token
+        will expose it to macro expansion, to the tune of much confusion and agony.
+
+    **************************************************************************************/
+    int last_errno;
```

Comment:
No, it should not be `extern int`: It is a part of the `PJ` object, so `extern`ing it would be a syntax error. 

The `PJ` local last_error is used in `proj_errno_set()`, and since I'm trying to introduce some slightly more sane, but backwards compatible and not very intrusive, error reporting mechanism for the OBS API. I think adding an `int` (i.e. typically 4 or 8 bytes) to the already rather large `PJ` object is a small price to pay.

I  see now that `proj_errno` lacks a check for the PJ local `last_error`, which I will add in a moment - the point is that it should return the local if nonzero, otherwise go on and search for the context specific, and (perhaps) finally the default context specific, because not all corners of the code can be supposed to support the PJ local reporting at once.

Note that the PJ local errno bubbles up to the context level (se comment in code), for compatibility with `proj_api.h` and `projects.h` based code.